### ISSUE TITLE
Failed to use "kill -HUP" to reload config when adding a new export

### ISSUE THREAD

**Issue by ColinHyu** [nfs-ganesha/nfs-ganesha#644]
Hi,

We use NFS-ganesha with CEPH FSAL and we are trying V4-dev.39 recently. (CEPH version is 13.2.4 mimic)

Every export has its config file and all files are included in /etc/ganesha/ganesha.conf.
```
%include /etc/ganesha/nfs_share.conf
```
And each config file contains
```
EXPORT
{
    SecType = sys;
    Anonymous_gid = 65534;
    Pseudo = /nfs_share;
    Squash = None;
    Access_Type = RW;
    FSAL
    {
        User_Id = cluster01;
        Name = CEPH;
    }
    Anonymous_uid = 65534;
    Path = /QSGW/cluster01/nfs_share;
    Export_Id = 1001;
    Transports = UDP,TCP;
    Protocols = 3,4;
}

NFSV4
{
    Only_Numeric_Owners = true;
}

NFS_KRB5
{
    KeytabPath = /etc/krb5.keytab;
    Active_krb5 = YES;
}

NFS_CORE_PARAM
{
    mount_path_pseudo = true;
}

LOG
{
    COMPONENTS
    {
        CONFIG = MID_DEBUG;
        EXPORT = MID_DEBUG;
        INODE = FULL_DEBUG;
    }
}
```
We use command `kill -HUP $(pidof ganesha.nfsd)` to reload configuration after we add, delete or modify any config file.

It can reload correctly after modifying or deleting a config file.

But adding config file is always failed. (if we restart ganesha process, everything is well.... )

For example, we already have started ganesha.nfsd, then we add a new export 'nfs_share' and reload.

The log is：
```
=~=~=~=~=~=~=~=~=~=~=~= PuTTY log 2020.10.30 15:48:52 =~=~=~=~=~=~=~=~=~=~=~=
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] log_conf_commit :CONFIG :EVENT :Switching to new component log levels
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :CONFIG :INFO :Reread exports starting
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] export_display :EXPORT :M_DBG :DEFAULTS 0x7f561c002ad8 Export 1 pseudo ((null)) with path ((null)) and tag ((null)) perms (options=03303002/00000000               ,     ,    ,               ,               ,         ,                ,                ,                )
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] init_export_root :EXPORT :DEBUG :About to lookup_path for ExportId=1001 Path=/QSGW/cluster01/nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_find_keyed_reason :INODE :F_DBG :Looking for hk=421aa12bf2a299 fsal=0x7f56c2b6e3e0 key=0x6d1a000000010000feffffffffffffff0000000000000000e903
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_new_entry :INODE :DEBUG :Adding a DIRECTORY, entry=0x7f561c15f2d0 setting MDCACHE_TRUST_CONTENT clearing MDCACHE_DIR_POPULATED
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_new_entry :INODE :F_DBG :New entry 0x7f561c15f2d0 added with fh_hk.key hk=421aa12bf2a299 fsal=0x7f56c2b6e3e0 key=0x6d1a000000010000feffffffffffffff0000000000000000e903
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_lookup_path :INODE :F_DBG :lookup_path Created entry 0x7f561c15f2d0 FSAL Ceph
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] init_export_root :EXPORT :DEBUG :Added root obj 0x7f561c15f308 FSAL MDCACHE for path /QSGW/cluster01/nfs_share on export_id=1001
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Looking for export for 
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share Rest nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_refresh_attrs :INODE :F_DBG :attrs  obj attributes Request Mask=0011dfce Valid Mask=0011dfce Supported Mask=0011dfce DIRECTORY numlinks=0x2 size=0x0 mode=0755 owner=0x0 group=0x0 atime=30/10/2020 04:18:25 mtime=30/10/2020 04:18:25
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_getattrs :INODE :F_DBG :attrs  obj attributes Request Mask=00000680 Valid Mask=0011dfce Supported Mask=0011dfce DIRECTORY numlinks=0x2 size=0x0 mode=0755 owner=0x0 group=0x0 atime=30/10/2020 04:18:25 mtime=30/10/2020 04:18:25
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_lookup :INODE :F_DBG :Lookup nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_try_get_cached :INODE :F_DBG :Look in cache nfs_share, trust content yes
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_avl_lookup :INODE :F_DBG :Lookup nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_avl_lookup :INODE :F_DBG :entry not found nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] trust_negative_cache :INODE :F_DBG :Entry 0x455510 Don't Trust negative cache
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_try_get_cached :INODE :F_DBG :mdcache_avl_lookup nfs_share failed trust negative no
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] trust_negative_cache :INODE :F_DBG :Entry 0x455510 Don't Trust negative cache
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_lookup :INODE :F_DBG :Try again nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_try_get_cached :INODE :F_DBG :Look in cache nfs_share, trust content yes
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_avl_lookup :INODE :F_DBG :Lookup nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_avl_lookup :INODE :F_DBG :entry not found nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] trust_negative_cache :INODE :F_DBG :Entry 0x455510 Don't Trust negative cache
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_try_get_cached :INODE :F_DBG :mdcache_avl_lookup nfs_share failed trust negative no
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] trust_negative_cache :INODE :F_DBG :Entry 0x455510 Don't Trust negative cache
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_lookup :INODE :DEBUG :Cache Miss detected for nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] lookup :EXPORT :DEBUG :PseudoFS LOOKUP of nfs_share may have failed due to export update
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdc_lookup_uncached :INODE :DEBUG :lookup nfs_share failed with File busy, retry
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] make_pseudofs_node :EXPORT :CRIT :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share LOOKUP nfs_share failed with File busy, retry
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] free_export_resources :EXPORT :DEBUG :Free resources for export 0x7f561c002ad8 id 1001 path /QSGW/cluster01/nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] free_export_resources :EXPORT :DEBUG :Export root 0x7f561c15f308
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] release_export :EXPORT :DEBUG :Released root obj 0x7f561c15f308 for path /QSGW/cluster01/nfs_share, pseudo /nfs_share on export_id=1001
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_unexport :EXPORT :DEBUG :About to unmap root entry 0x7f561c15f2d0 and possibly free it for export 1001 path /QSGW/cluster01/nfs_share pseudo /nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] cih_remove_latched :INODE :F_DBG :Unhashing entry 0x7f561c15f2d0
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] _mdcache_lru_unref :INODE :DEBUG :LRU_ENTRY_CLEANUP of entry 0x7f561c15f2d0
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_lru_clean :INODE :F_DBG :Trusting op_ctx export id 1001
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] mdcache_dirent_invalidate_all :INODE :F_DBG :Invalidating directory for 0x7f561c15f2d0, clearing MDCACHE_DIR_POPULATED setting MDCACHE_TRUST_CONTENT and MDCACHE_TRUST_DIR_CHUNKS
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] free_export_resources :EXPORT :DEBUG :release_export complete
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] free_export_resources :EXPORT :DEBUG :Goodbye export 0x7f561c002ad8 path /QSGW/cluster01/nfs_share pseudo /nfs_share
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :EXPORT :DEBUG :Exports before update
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :EXPORT :DEBUG :Export     0 pseudo (/) with path (/) and tag ((null)) perms (options=0221f080/0771f3e7 no_root_squash, --r-, -4-, ---, TCP, ----,               ,         ,                ,                ,                , none, sys, krb5, krb5i, krb5p)
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] prune_pseudofs_subtree :EXPORT :DEBUG :Exxport 0 pseudo / export gen 18446744073709551615 current gen 6 prune unmount no ancestor_is_defunct no
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] prune_pseudofs_subtree :EXPORT :DEBUG :Export 0 Pseudo / not unmounted
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :CONFIG :EVENT :Reread exports complete
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :EXPORT :INFO :Exports after update
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] reread_exports :EXPORT :INFO :Export     0 pseudo (/) with path (/) and tag ((null)) perms (options=0221f080/0771f3e7 no_root_squash, --r-, -4-, ---, TCP, ----,               ,         ,                ,                ,                , none, sys, krb5, krb5i, krb5p)
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/nfs_share.conf:1): 1 validation errors in block EXPORT
30/10/2020 07:48:55 : epoch 5f9b9411 : joneschan : ganesha.nfsd-7505[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/nfs_share.conf:1): Errors processing block (EXPORT)
```
According log, we think maybe something wrong in mdcache_avl_up due to "entry not found nfs_share" message, but we're not sure about that.

Any ideas?

**Comment by dang** [comment]
So, the "entry not found" message from mdcache_avl_lookup() is not, itself, a problem, as it just indicates a cache miss.   The important one is:
mdc_lookup_uncached :INODE :DEBUG :lookup nfs_share failed with File busy, retry

This means that, when the lookup() method was called on the FSAL_CEPH returned EBUSY from libcephfs.

**Comment by ColinHyu** [comment]
Hi, deng 

Thanks for your reply.

According your reply, we enable FSAL full_debug and put our logs in the lookup() of FSAL_CEPH.
```
rc = fsal_ceph_ll_lookup(export->cmount, dir->i, path, &i, &stx,!!attrs_out, &op_ctx->creds);
LogFullDebug(COMPONENT_FSAL, "Lookup end with %d", rc);
```
We think the subcall in mdc_lookup_uncached calls ceph_fsal_lookup.

However, it shows up only once when we restart ganesha.
```
03/11/2020 07:27:10 : epoch 5fa105ba : joneschan : ganesha.nfsd-13757[sigmgr] ceph_fsal_lookup :FSAL :F_DBG :Lookup ..
03/11/2020 07:27:10 : epoch 5fa105ba : joneschan : ganesha.nfsd-13757[sigmgr] ceph_fsal_lookup :FSAL :F_DBG :Lookup end with 0
```
but not before "mdc_lookup_uncached :INODE :DEBUG :lookup nfs_share failed with File busy, retry" message
```
03/11/2020 07:27:10 : epoch 5fa105ba : joneschan : ganesha.nfsd-13757[sigmgr] lookup :EXPORT :DEBUG :PseudoFS LOOKUP of nfs_share may have failed due to export update
03/11/2020 07:27:10 : epoch 5fa105ba : joneschan : ganesha.nfsd-13757[sigmgr] mdc_lookup_uncached :INODE :DEBUG :lookup nfs_share failed with File busy, retry
```
We also found that the lookup() in fsal_pseudo can return ERR_FSAL_DELAY, which means "File busy, retry" too.

So maybe the subcall in mdc_lookup_uncached calls the lookup() in fsal_pseudo but not ceph_fsal_lookup？


Besides, we found a log when we start Ganesha
```
03/11/2020 07:27:10 : epoch 5fa105ba : joneschan : ganesha.nfsd-13757[sigmgr] register_callbacks :FSAL :WARN :This libcephfs does not support registering callbacks. Ganesha will be unable to respond to MDS cache pressure.
```
Is it a possible reason for this error?

**Comment by dang** [comment]
The warning from libcephfs is fine, it just means that you have an older version, so some features will not be active.  Basic functionality works fine.

It does look like the path is being looked up in PSEUDO, which shouldn't happen, because that is a junction into CEPH.  Odd.  You can tell for sure by turning on DEBUG for EXPORT, which would cause PSEUDO to log a message in that case.

@ffilz any ideas?

**Comment by ffilz** [comment]
Where do you see the lookup of the path in Pseudo?

I agree, running with the EXPORT component on DEBUG would help, you can put this in your config, since we're debugging an export reload, let's go ahead and do FULL_DEBUG. The log config is also reloaded (before exports) on SIGHUP.

LOG {
    COMPONENTS {
        EXPORT = FULL_DEBUG;
    }
}

**Comment by ColinHyu** [comment]
Hi,

I see the lookup() of Pseudo in \src\FSAL\FSAL_PSEUDO\hanedle.c
and there is an "out" label 
```
out:

	if (error == ERR_FSAL_NOENT && is_export_update_in_progress()) {
		/* An export update may be the cause of the failure. Tell the
		 * client to retry.
		 */
		LogDebug(COMPONENT_EXPORT,
			 "PseudoFS LOOKUP of %s may have failed due to export update",
			 path);

		error = ERR_FSAL_DELAY;
	}
```
And the log of "EXPORT = FULL_DEBUG" is
```
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] log_conf_commit :CONFIG :EVENT :Switching to new component log levels
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] alloc_export :EXPORT :F_DBG :Allocated export 0x7f0e680070d8
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_init :EXPORT :F_DBG :Allocated export 0x7f0e680070d8
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_display :EXPORT :M_DBG :DEFAULTS 0x7f0e680070d8 Export 1 pseudo ((null)) with path ((null)) and tag ((null)) perms (options=03303002/00000000               ,     ,    ,               ,               ,               ,                ,                ,                )
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] fsal_cfg_commit :EXPORT :F_DBG :get export ref for id 1001 /QSGW/cluster01/nfs_share, refcount = 2
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] clean_export_paths :EXPORT :F_DBG :Cleaning paths for 1001 fullpath/QSGW/cluster01/nfs_share pseudopath /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] clean_export_paths :EXPORT :F_DBG :Final paths for 1001 fullpath /QSGW/cluster01/nfs_share pseudopath /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] register_callbacks :FSAL :WARN :This libcephfs does not support registering callbacks. Ganesha will be unable to respond to MDS cache pressure.
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] clear_op_context_export_impl :EXPORT :F_DBG :put export ref for id 1001 /QSGW/cluster01/nfs_share, refcount = 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_commit_common :EXPORT :F_DBG :Processing 0x7f0e680070d8
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Searching for exportmatching pseudo path /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Comparing /nfs_share 10 to / 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_path_locked :EXPORT :F_DBG :Searching for export matching path /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] init_export_root :EXPORT :F_DBG :get export ref for id 1001 /QSGW/cluster01/nfs_share, refcount = 2
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] init_export_root :EXPORT :DEBUG :About to lookup_path for ExportId=1001 Path=/QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] init_export_root :EXPORT :DEBUG :Added root obj 0x7f0e6815e608 FSAL MDCACHE for path /QSGW/cluster01/nfs_share on export_id=1001
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] clear_op_context_export_impl :EXPORT :F_DBG :put export ref for id 1001 /QSGW/cluster01/nfs_share, refcount = 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Looking for export for
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Searching for export matching pseudo path
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Comparing  0 to / 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :get export ref for id 0 /, refcount = 2
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] pseudo_mount_export :EXPORT :DEBUG :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share Rest nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] lookup :EXPORT :DEBUG :PseudoFS LOOKUP of nfs_share may have failed due to export update
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] make_pseudofs_node :EXPORT :CRIT :BUILDING PSEUDOFS: Export_Id 1001 Path /QSGW/cluster01/nfs_share Pseudo Path /nfs_share LOOKUP nfs_share failed with File busy, retry
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] clear_op_context_export_impl :EXPORT :F_DBG :put export ref for id 0 /, refcount = 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_init :EXPORT :F_DBG :Releasing export 0x7f0e680070d8
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_init :EXPORT :F_DBG :put export ref for id 1001 /QSGW/cluster01/nfs_share, refcount = 0
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] free_export_resources :EXPORT :DEBUG :Free resources for export 0x7f0e680070d8 id 1001 path /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] free_export_resources :EXPORT :DEBUG :Export root 0x7f0e6815e608
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] release_export :EXPORT :DEBUG :Released root obj 0x7f0e6815e608 for path /QSGW/cluster01/nfs_share, pseudo /nfs_share on export_id=1001
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] release_export :EXPORT :F_DBG :About to unexport from FSAL root obj 0x7f0e6815e608 for path /QSGW/cluster01/nfs_share, pseudo /nfs_share on export_id=1001
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] mdcache_unexport :EXPORT :DEBUG :About to unmap root entry 0x7f0e6815e5d0 and possibly free it for export 1001 path /QSGW/cluster01/nfs_share pseudo /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] mdcache_unexport :EXPORT :F_DBG :Disposing of entry 0x7f0e6815e5d0
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] release_export :EXPORT :F_DBG :About to put_ref root obj 0x7f0e6815e608 for path /QSGW/cluster01/nfs_share, pseudo /nfs_share on export_id=1001
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] free_export_resources :EXPORT :DEBUG :release_export complete
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] free_export_resources :EXPORT :DEBUG :Goodbye export 0x7f0e680070d8 path /QSGW/cluster01/nfs_share pseudo /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /QSGW/cluster01/nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] gsh_refstr_release :EXPORT :F_DBG :Releasing refstr /nfs_share
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] alloc_export :EXPORT :F_DBG :Allocated export 0x7f0e68041b58
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] export_init :EXPORT :F_DBG :Allocated export 0x7f0e68041b58
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] reread_exports :EXPORT :DEBUG :Exports before update
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] reread_exports :EXPORT :DEBUG :Export 0 pseudo (/) with path (/) and tag ((null)) perms (options=0221f080/0771f3e7 no_root_squash, --r-, -4-, ---, TCP, ----,               ,         ,                ,                      ,                , none, sys, krb5, krb5i, krb5p)
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Searching for export matching pseudo path /
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :Comparing / 1 to / 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] get_gsh_export_by_pseudo_locked :EXPORT :F_DBG :get export ref for id 0 /, refcount = 2
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] prune_pseudofs_subtree :EXPORT :DEBUG :Exxport 0 pseudo / export gen 18446744073709551615 current gen 2 prune unmount no ancestor_is_defunct no
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] prune_pseudofs_subtree :EXPORT :DEBUG :Export 0 Pseudo / not unmounted
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] prune_pseudofs_subtree :EXPORT :F_DBG :put export ref for id 0 /, refcount = 1
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] reread_exports :CONFIG :EVENT :Reread exports complete
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] reread_exports :EXPORT :INFO :Exports after update
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] reread_exports :EXPORT :INFO :Export  0 pseudo (/) with path (/) and tag ((null)) perms (options=0221f080/0771f3e7 no_root_squash, --r-, -4-, ---, TCP, ----,               ,         ,                ,                      ,                , none, sys, krb5, krb5i, krb5p)
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/share/CACHEDEV1_DATA/.config/nfs/ganesha.conf.d/nfs_share.conf:1): 1 validation errors in block EXPORT
04/11/2020 02:16:54 : epoch 5fa20eb7 : joneschan : ganesha.nfsd-26420[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/share/CACHEDEV1_DATA/.config/nfs/ganesha.conf.d/nfs_share.conf:1): Errors processing block (EXPORT)
```

**Comment by ffilz** [comment]
Ok, got it, will have a fix ready shortly.

**Comment by ffilz** [comment]
Fix in https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/505453

**Comment by ffilz** [comment]
BTW, I didn't catch this in my testing of reloading exports because all the exports I added in my testing were sub-exports of FSAL_VFS exports... This only happens when the new export is parented by the PseudoFS... Which of course is where MOST people add new exports.... OOPS testing fail... :-)

**Comment by ColinHyu** [comment]
This patch works ! Thanks !

Now we can reload correctly after we add a new export !!

But there is another problem,

When reloading is finish, we found that the "showmount -e" result in nfs client is
```
root@NFS-CLIENT:/mnt# showmount -e 10.19.60.65
Export list for 10.19.60.65:
/QSGW/cluster01/nfs_share  (everyone)
/QSGW/cluster01/nfs_share2 (everyone)
```
but if we "restart" ganesha.nfsd the result becomes
```
root@NFS-CLIENT:/mnt# showmount -e 10.19.60.65
Export list for 10.19.60.65:
/nfs_share  (everyone)
/nfs_share2 (everyone)
```
The second result is what we expected

It looks like "mount_path_pseudo" not working when reloading ?

**Comment by ffilz** [comment]
Could you open a new issue for that?

**Comment by ffilz** [comment]
And then we can close this issue.

**Comment by ColinHyu** [comment]
OK!

**Comment by ColinHyu** [comment]
Hi,

I have another environment with ganesha-v2.8-stable (also FSAL_CEPH)

Could I patch https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/505453 in ganesha-v2.8-stable version, too?

**Comment by ffilz** [comment]
https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/505453 only applies to next, it is fixing the changes made to export reload earlier in the V4-dev series.

Issue #647 is something different.

**Comment by ColinHyu** [comment]
OK! Thanks!

I will open a new issue for nfs-ganesha-2.8 stable if necessary.

**Comment by ffilz** [comment]
Closing as resolved.
